client/allocrunner/tasklifecycle/gate.go

```diff
@@ -0,0 +1,80 @@
+package tasklifecycle
+
+const (
+	gateClosed = false
+	gateOpened = true
+)
+
+// Gate is used by the Coordinator to block or allow tasks from running.
+//
+// It provides a channel that taskRunners listens on to determine when they are
+// allowed to run. The Gate has an infinite loop that is either feeding this
+// channel (therefore allowing listeners to proceed) or not doing anything
+// (causing listeners to block an wait).
+//
+// The Coordinator uses the Gate Open() and Close() methods to control this
+// producer loop.
+type Gate struct {
+	sendCh     chan struct{}
+	updateCh   chan bool
+	shutdownCh <-chan struct{}
+}
+
+// NewGate returns a new Gate that is initially closed. The Gate should not be
+// used after the shutdownCh is closed.
+func NewGate(shutdownCh <-chan struct{}) *Gate {
+	g := &Gate{
+		sendCh:     make(chan struct{}),
+		updateCh:   make(chan bool),
+		shutdownCh: shutdownCh,
+	}
+	go g.run(gateClosed)
+
+	return g
+}
+
+// WaitCh returns a channel that the listener must block on before starting its
+// task.
+func (g *Gate) WaitCh() <-chan struct{} {
+	return g.sendCh
+}
+
+// Open is used to allow listeners to proceed.
+func (g *Gate) Open() {
+	select {
+	case <-g.shutdownCh:
+	case g.updateCh <- gateOpened:
+	}
+}
+
+// Close is used to block listeners from proceeding.
+func (g *Gate) Close() {
+	select {
+	case <-g.shutdownCh:
+	case g.updateCh <- gateClosed:
+	}
+}
+
+// run starts the infinite loop that feeds the channel if the Gate is opened.
+func (g *Gate) run(initState bool) {
+	isOpen := initState
```

Comment:
I think this needs to close sendCh when it exits so that waiters on `WaitCh()` don't block indefinitely?
```suggestion
func (g *Gate) run(initState bool) {
	defer close(g.sendCh)

	isOpen := initState
```

Reply:
Good question, I'm not sure what's the best behaviour here. Closing the channel will leave the gate open on exit, which may result in tasks unexpectedly running? 

So I thought that preserving the gate state would be the safer option, but this could cause goroutines to be blocked forever and leak resources?

Reply:
Yeah, definitely want to avoid starting tasks up due to an agent shutdown signal!

Regardless of the behavior here we need to ensure _callers do the right thing:_ callers must always check for an shutdown condition to determine whether to transition states or just escape from the runner ASAP because it's shutting down.

So I think it's actually best to make it explicitly the *callers* problem to figure out what's happening:

1. If WaitCh() is closed: transition states
2. If shutdownCh is closed: shutdown

But the comment on WaitCh should probably make it clear that handling shutdowns are the callers concern so no one ends up blocking forever.

Reply:
Ahh OK. I have this in the `NewGate` docs, which is `shutdownCh` is passed:

> The Gate should not be used after the shutdownCh is closed.

But I will add a similar not to the `WaitCh` as well 👍

Reply:
Added the following comment:

```go
// Callers must also check the state of the shutdownCh used to create the Gate
// to avoid blocking indefinitely.
```